pkg/authorization/authorizer/cache/authorizer.go

```diff
@@ -0,0 +1,110 @@
+package cache
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/hashicorp/golang-lru"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	kerrs "k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/util"
+	"k8s.io/kubernetes/pkg/util/sets"
+
+	"github.com/openshift/origin/pkg/authorization/authorizer"
+)
+
+type CacheAuthorizer struct {
+	authorizer authorizer.Authorizer
+
+	cache *lru.Cache
+
+	maxAge time.Duration
+	now    func() time.Time
+}
+
+type cacheRecord struct {
+	created time.Time
+	allowed bool
+	reason  string
+	err     error
+}
+
+// NewAuthorizer returns an authorizer that caches the results of the given authorizer
+func NewAuthorizer(a authorizer.Authorizer, maxAge time.Duration, maxCount int) (authorizer.Authorizer, error) {
+	cache, err := lru.New(maxCount)
+	if err != nil {
+		return nil, err
+	}
+	return &CacheAuthorizer{
+		authorizer: a,
+		cache:      cache,
+		maxAge:     maxAge,
+		now:        time.Now,
+	}, nil
+}
+
+func (c *CacheAuthorizer) Authorize(ctx kapi.Context, a authorizer.AuthorizationAttributes) (allowed bool, reason string, err error) {
+	key, err := cacheKey(ctx, a)
+	if err != nil {
+		glog.Infof("could not build cache key for %#v: %v", a, err)
+		return c.authorizer.Authorize(ctx, a)
+	}
+
+	if value, hit := c.cache.Get(key); hit {
+		switch record := value.(type) {
+		case *cacheRecord:
+			if record.created.Add(c.maxAge).After(c.now()) {
+				return record.allowed, record.reason, nil
+			} else {
+				glog.Infof("cache record expired for %s", key)
+			}
+		default:
+			util.HandleError(fmt.Errorf("invalid cache record type for key %s: %#v", key, record))
+		}
+	}
+
+	allowed, reason, err = c.authorizer.Authorize(ctx, a)
+
+	// Don't cache results if there was an error unrelated to authorization
+	if err == nil || kerrs.IsForbidden(err) {
+		c.cache.Add(key, &cacheRecord{created: c.now(), allowed: allowed, reason: reason, err: err})
+	}
+
+	return allowed, reason, err
+}
+
+func (c *CacheAuthorizer) GetAllowedSubjects(ctx kapi.Context, attributes authorizer.AuthorizationAttributes) (sets.String, sets.String, error) {
+	return nil, nil, errors.New("unsupported")
+}
+
+func cacheKey(ctx kapi.Context, a authorizer.AuthorizationAttributes) (string, error) {
+	if a.GetRequestAttributes() != nil {
+		return "", errors.New("cannot cache request attributes")
+	}
+
+	keyData := map[string]interface{}{
```

Comment:
This is prone to slippage as APIs drift.  I don't want reflective madness inside of the runtime code, but how about a reflect test that checks for keys for all available `Get*` functions from the `AuthorizationAttributes`?